contracts/Vault.sol

```diff
@@ -123,21 +130,27 @@ contract Vault is ExchangeRate, Initializable, OwnableUpgradeSafe {
             aTokenAmount = balance;
         }
 
+        require(aTokenAmount > 0, "amount to withdraw cannot be zero");
+        bToken.burn(address(this), balance);
+        require(aTokenAmount >= depositorToPrincipal[msg.sender], "atoken must be more than depositorPrincipal");
+        require(depositorReserve >= SafeMath.sub(aTokenAmount, depositorToPrincipal[msg.sender]), "atoken must be more than depositorPrincipal");
         depositorReserve = SafeMath.sub(
-            depositorReserve, 
+            depositorReserve,
             SafeMath.sub(aTokenAmount, depositorToPrincipal[msg.sender])
             );
+        require(principal >= depositorToPrincipal[msg.sender], "principal must be over depositorPrincipal");
         principal = SafeMath.sub(principal, depositorToPrincipal[msg.sender]);
         depositorToPrincipal[msg.sender] = 0;
         // withdraw atokens
-        bToken.burn(address(this), balance);
         aToken.approve(address(aavePool), aTokenAmount);
-        aavePool.withdraw(
+        uint256 withdrawn = aavePool.withdraw(
             address(token),
             aTokenAmount,
             msg.sender
         );
-        _updateCheckpointInterest();
+        if (startedSurplus) {
```

Comment:
seeing several places where we have:
```
 if(startedSurplus) {
       _updateCheckpointInterest();
}
```

that smells to me like the` if (startedSurplus)` check belongs inside the `_updateCheckpointInterest()` function, or? less duplicated code and if we have add  another _updateCheckpointInterest() somewhere, we might forget this. (not sure about smart contract optimization, but i don't think this will use more gas during runtime. not sure though.

Reply:
that's right, _updateCheckpointInterest() only happens after surplus has started. I will move it into it's own function to keep our code DRY.